syntax = "proto3";

package voicebot;

service VoiceBot {
    // Bidirectional streaming for real-time audio and control
    rpc StreamCall (stream CallEvent) returns (stream CallAction);
}

enum Codec {
    PCMU = 0;
    PCMA = 8;
    PCM16 = 10; // Raw linear PCM
}

enum Direction {
    UPLINK = 0;   // User -> Bot
    DOWNLINK = 1; // Bot -> User
}

message CallEvent {
    string call_id = 1;

    oneof payload {
        Config config = 2;       // Sent once at start
        AudioChunk audio = 3;    // Recurring audio packets
        DtmfEvent dtmf = 4;      // Detected DTMF
        ControlEvent control = 5;// Hangup, etc.
    }
}

message Config {
    int32 sample_rate = 1;
    Codec codec = 2;
}

message AudioChunk {
    uint64 seq = 1;
    uint64 timestamp = 2;
    bytes data = 3;
}

message DtmfEvent {
    string digit = 1; // "0"-"9", "*", "#", "A"-"D"
    bool end = 2;
}

message ControlEvent {
    enum Type {
        HANGUP = 0;
    }
    Type type = 1;
}

// Actions from Bot -> Gateway
message CallAction {
    oneof payload {
        AudioChunk audio = 1;    // Audio to play back to caller
        BotControl control = 2; // Hangup, Transfer, etc.
    }
}

message BotControl {
    enum Type {
        HANGUP = 0;
        TRANSFER = 1;
    }
    Type type = 1;
    string transfer_target = 2; // For REFER
}
